### PR TITLE
[cn0552 p1] ad7746 driver updates

### PR DESCRIPTION
**Pull request by buha**
AD7746 driver updates

Requires #1006 to be merged to work.

### CODE REVIEW

File: projects/cn0552/src/app/headless.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+#include <sys/platform.h>
+#include "adi_initialize.h"
+#include <drivers/pwr/adi_pwr.h>
+
+#include "iio.h"
+#include "iio_ad7746.h"
+#include "iio_app.h"
+#include "ad7746.h"
+#include "irq.h"
```
Comment (amiclaus) [inline projects/cn0552/src/app/headless.c:51]: I believe you don't need all these includes like irq, irq_extra, uart, uart_extra if you are using iio_app.

**Review by amiclaus — COMMENTED** [review]
General comment:
In my opinion there are too many commits for one Pull Request.
I would do separate Pull Requests and add them progressively.
1. AD7746 fixes
2. IIO updates
3. AD7746 IIO driver + CN0552 project.

File: projects/cn0552/src/app/headless.c (hunk 1 above)
Comment (buha) [reply projects/cn0552/src/app/headless.c:51]: yeah i agree about too many commits, there many changes
will see about those includes, maybe i can cut some out

File: drivers/adc/ad7746/ad7746.h [hunk 2]
```diff
@@ -182,6 +182,8 @@ struct ad7746_init_param {
 struct ad7746_dev {
 	i2c_desc *i2c_dev;
 	uint8_t buf[AD7746_NUM_REGISTERS + 1u];
+	uint8_t capdac[2][2];
```
Comment (mchindri) [inline drivers/adc/ad7746/ad7746.h:185]: Can 2 and 2 be a define?

File: include/util.h [hunk 3]
```diff
@@ -111,6 +111,8 @@
 
 #define UNUSED_PARAM(x) ((void)x)
 
+#define shift_right(x, s) ((x) < 0 ? -(-(x) >> (s)) : (x) >> (s))
```
Comment (mchindri) [inline include/util.h:114]: If the variable is signed, isn't this done by default?

File: libraries/iio/iio.c [hunk 4]
```diff
@@ -284,27 +293,6 @@ static ssize_t iio_phy_write(const char *buf, size_t len)
 	return -EINVAL;
 }
 
-/* Get string for channel id from channel type */
-static char *get_channel_id(enum iio_chan_type type)
```
Comment (mchindri) [inline libraries/iio/iio.c:288]: Function using get_channel_id should be updated in this commit

File: libraries/iio/iio_types.h [hunk 5]
```diff
@@ -51,6 +51,18 @@
 /*************************** Types Declarations *******************************/
 /******************************************************************************/
 
+
+enum iio_val {
```
Comment (mchindri) [inline libraries/iio/iio_types.h:55]: Where is this used?
It seems not related to this commit

File: libraries/iio/iio.c [hunk 6]
```diff
@@ -296,19 +296,19 @@ static ssize_t iio_phy_write(const char *buf, size_t len)
...
+				sprintf(buff, "%s%d-%s%d", iio_chan_type_string[ch->ch_type],
+					(int)ch->channel, iio_chan_type_string[ch->ch_type],
 					(int)ch->channel2);
 			else
-				sprintf(buff, "%s%d", get_channel_id(ch->ch_type),
+				sprintf(buff, "%s%d", iio_chan_type_string[ch->ch_type],
 					(int)ch->channel);
 		} else {
-			sprintf(buff, "%s", get_channel_id(ch->ch_type));
+			sprintf(buff, "%s", iio_chan_type_string[ch->ch_type]);
```
Comment (mchindri) [inline libraries/iio/iio.c:311]: Below changes are related with a previous commit

File: libraries/iio/iio_types.h [hunk 7]
```diff
@@ -161,6 +167,8 @@ struct iio_channel {
 	 *  for angular rate when applied to channel2 will make make the
 	 *  IIO_ANGL_VEL have anglvel_x which corresponds to the x-axis. */
 	int 			channel2;
+	/** Driver specific identifier. */
+	unsigned long		address;
```
Comment (mchindri) [inline libraries/iio/iio_types.h:171]: Is the linux naming?
Doesn't it make more sense to be void *context or something like that?

File: include/util.h (hunk 3 above)
Comment (buha) [reply include/util.h:114]: no, nothing guarantees you that

File: libraries/iio/iio_types.h (hunk 7 above)
Comment (buha) [reply libraries/iio/iio_types.h:171]: i don't think changing names will improve anything - it will actually make it worse, every time you wanna port an iio driver implementation between linux <-> no-OS you need to think about which corresponds to which

File: libraries/iio/iio.c (hunk 4 above)
Comment (buha) [reply libraries/iio/iio.c:288]: of course, i did it in a different commit, will try to move changes around

**Comment by buha** [comment]
V2:
- implemented feedback above (moved commits around, remove useless includes)

**Comment by buha** [comment]
V3:
- split into multiple PRs

File: include/util.h (hunk 3 above)
Comment (mchindri) [reply include/util.h]: It seems to be compiler dependent (https://en.wikipedia.org/wiki/Bitwise_operations_in_C#Right_shift_%3E%3E) 
Good to know :)

File: drivers/adc/ad7746/ad7746.c [hunk 8]
```diff
@@ -504,5 +509,49 @@ int32_t ad7746_get_cap_data(struct ad7746_dev *dev, uint32_t *cap_data)
...
+	    md != AD7746_MODE_GAIN_CALIB)
+		return -EINVAL;
+
+	c->md = md;
+	ret = ad7746_set_config(dev, *c);
+	if (ret < 0)
+		return ret;
+
+	do {
+		mdelay(20);
```
Comment (amiclaus) [inline drivers/adc/ad7746/ad7746.c:545]: is this a datasheet related value? or trial/error value? might fit a code comment here :)

Comment (buha) [reply drivers/adc/ad7746/ad7746.c:545]: yeah it needs a comment !

**Review by adrimbarean — APPROVED** [review]
Technically the last commit can go in the PR that will use it, but approve.

**Comment by buha** [comment]
V4: 
- add chip id
- comment that delay
- remove capdac stuff from here -> got moved to iio part of the driver

**Review by mchindri — APPROVED** [review]

**Review by dbogdan — APPROVED** [review]
